rust/src/meas/geometric/mod.tex

```diff
@@ -0,0 +1,283 @@
+
+\subsubsection*{Preconditions}
+To ensure the correctness of the output, we require the following preconditions:
+
+\begin{itemize}
+    \item \textbf{User-specified types:}
+    \begin{itemize}
+        \item Domain \texttt{AllDomain} must be of elements of type \texttt{T}
+        \item Variable \texttt{scale} must be of type \texttt{QO}
+        \item Variable \texttt{bounds} must be of type \texttt{Option(D::Atom, D::Atom)}
+        \item Type \texttt{D} must have trait \texttt{GeometricDomain}
+        \item Type \texttt{D::Atom} must have traits \texttt{TotalOrd} and \texttt{InfCast<QO>}
+        \item Type \texttt{QO} must have traits \texttt{float} and \texttt{DistanceConstant<D::Atom>} 
+        \item Type \texttt{f64} must have trait \texttt{InfCast<QO>}
+        \item Type \texttt{T} must have traits \texttt{SampleTwoSidedGeometric} and \texttt{CheckNull}
+    \end{itemize}
+\end{itemize}
+
+\vicki{not sure about the domain/T stuff; it's mentioned in the implementation of the domain so I figured should also have it here} 
+
+\subsubsection*{Postconditions}
+
+\begin{itemize}
+    \item A \texttt{Measurement} is returned (i.e., if a \texttt{Measurement} cannot be returned successfully, then an error should be returned).
+\end{itemize}
+
+
+\begin{lstlisting}[language=Python,  escapechar=|]
+def make_base_geometric(scale, bounds={"lower": T_MIN, "upper": T_MAX}):
+    input_domain = AllDomain(T)
+    output_domain = AllDomain(T)
+    input_metric = L1Distance(T)
+    similarity_metric = MaxDivergence(QO)
+    
+    # check scale sign
+    if (scale < 0): |\label{line:scale}|
+        raise Exception("scale must not be negative")
+        
+    # check bounds - can discuss if python dict is best representation of this
+    if (bounds["lower"] > bounds["upper"]): |\label{line:bounds}|
+        raise Exception("lower may not be greater than upper")
+
+    def privacy_map(d_in: u_32) -> float: |\label{line:map}|
+        if (d_in < 0): |\label{line:positivedin}|
+            raise Exception("sensitivity must be non-negative")
+        if (scale == 0):
+            return QO_MAX
+        return d_in * inf_div(scale);
+
+    # these are two separate functions in original, but can discuss if we want to combine the two
+    def noise_function(arg : T, scale : QO, bounds : Optional[{T, T}]) -> Function<f64, Optional[{T, T}]>: |\label{line:noise_fn}|
+        return function(inf_cast(scale), bounds)
+        
+    def function(arg : T, scale : f64, bounds: Optional[{T, T}]) -> <D, QO>: |\label{line:fn}|
+        sample_two_sided_geometric(arg, scale, bounds)
```

Comment:
I don't think you should define this function inside the constructor! It has already been defined outside of the function because of the GeometricDomain trait bound on D. 
Maybe the syntax we should use to refer to it is `D.noise_function`? Then, outside the function, define what `D.noise_function` is when D is `VectorDomain<AllDomain<T>>` and when D is `AllDomain<T>`.

Then you can build the function as follows:
```python
function = D.noise_function(scale, bounds)
```

Happy to adjust naming in the code to make this more clear.